File: Top100/groupAnagrams.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> ans;
        vector<string> sorted;
        string temp;
        for (int i = 0; i < strs.size(); i++) {
            temp = strs[i];
            sort(temp.begin(), temp.end());
            
            int j = 0;
            for (j = 0; j < sorted.size(); j++){
                if (temp == sorted[j]) {
                    ans[j].push_back(strs[i]);
                    break;
                }
            }
            if (j == sorted.size()) {
                sorted.push_back(temp);
                ans.resize(j + 1); // resize 不会清除之前的数据
                ans[j].push_back(strs[i]);
            }    
        }
        return ans;
    }
};

// 竟然一下通过了！！

File: Top100/groupAnagrams.java
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.Arrays;

public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0)
            return new ArrayList<>();
        Map<String, List> ans = new HashMap<String, List>();
        for (String s : strs) {
            char[] ca = s.toCharArray();
            Arrays.sort(ca);
            String key = String.valueOf(ca); // String.valueOf(char[] data) : 将 char 数组 data 转换成字符串
            if (!ans.containsKey(key)) // 不包含
                ans.put(key, new ArrayList<>());
            ans.get(key).add(s);
        }
        return new ArrayList<>(ans.values());
    }
}
